packages/dara-core/dara/core/internal/encoder_registry.py

```diff
@@ -0,0 +1,69 @@
+"""
+Copyright 2023 Impulse Innovations Limited
+
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+
+from typing import Any, Callable, MutableMapping, Type, TypedDict
+
+import numpy
+import pandas
+
+
+class Encoder(TypedDict):
+    serialize: Callable
+    deserialize: Callable
+
+
+def _get_numpy_dtypes_encoder(typ: Type[Any]):
+    """
+    Construct numpy generic datatype
+
+    :param typ: The numpy generic datatype class
+    """
+    return Encoder(serialize=lambda x: x.item(), deserialize=lambda x: typ(x))
+
+
+# A encoder_registry to handle serialization/deserialization for numpy/pandas type
+encoder_registry: MutableMapping[Type[Any], Encoder] = {
```

Comment:
@tamara-causalens can you check whether this covers most use common data science use cases (for the open source). Internal type serialization can be added to our internal plugin

Reply:
> @tamara-causalens can you check whether this covers most use common data science use cases (for the open source). Internal type serialization can be added to our internal plugin

I handle all the subclass of generic type in Numpy. In pandas DataFrame, the datatype could a numpy dtype or pyarrow dtype. Let me know if we should handle the pyarrow as well.
```
<class 'numpy.number'>
     <class 'numpy.integer'>
         <class 'numpy.signedinteger'>
             <class 'numpy.int8'>
             <class 'numpy.int16'>
             <class 'numpy.int32'>
             <class 'numpy.int64'>
             <class 'numpy.longlong'>
             <class 'numpy.timedelta64'>
         <class 'numpy.unsignedinteger'>
             <class 'numpy.uint8'>
             <class 'numpy.uint16'>
             <class 'numpy.uint32'>
             <class 'numpy.uint64'>
             <class 'numpy.ulonglong'>
     <class 'numpy.inexact'>
         <class 'numpy.floating'>
             <class 'numpy.float16'>
             <class 'numpy.float32'>
             <class 'numpy.float64'>
             <class 'numpy.longdouble'>
         <class 'numpy.complexfloating'>
             <class 'numpy.complex64'>
             <class 'numpy.complex128'>
             <class 'numpy.clongdouble'>
 <class 'numpy.flexible'>
     <class 'numpy.character'>
         <class 'numpy.bytes_'>
         <class 'numpy.str_'>
     <class 'numpy.void'>
         <class 'numpy.record'>
 <class 'numpy.bool_'>
 <class 'numpy.datetime64'>
 <class 'numpy.object_'>

```